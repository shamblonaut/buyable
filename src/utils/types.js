import PropTypes from "prop-types";

export const numberBetween = (min, max) => {
  return (props, propName, componentName) => {
    const value = props[propName];

    // isRequired
    if (value == null) {
      return new Error(
        `The prop \`${propName}\` is marked as required in \`${componentName}\`, but its value is \`${value}\`.`,
      );
    }

    // number
    if (typeof value !== "number") {
      return new Error(
        `Invalid prop \`${propName}\` of type \`${typeof value}\` supplied to \`${componentName}\`, expected \`number\`.`,
      );
    }

    // between min and max
    if (value < min || value > max) {
      return new Error(
        `Invalid prop \`${propName}\` of value \`${value}\` supplied to \`${componentName}\`, expected a number between ${min} and ${max}.`,
      );
    }

    // Passed validation
    return null;
  };
};

export const ProductType = PropTypes.exact({
  id: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  category: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
  rating: PropTypes.exact({
    rate: numberBetween(0, 5),
    count: PropTypes.number.isRequired,
  }).isRequired,
});
