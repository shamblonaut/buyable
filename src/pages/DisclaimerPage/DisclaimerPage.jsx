import { Page, Heading, Disclaimer } from "./DisclaimerPage.styles";

const DisclaimerPage = () => (
  <Page>
    <Heading>Disclaimer</Heading>
    <Disclaimer>
      Sorry to disappoint you, but this was a fake online store that I built to
      practice React development. Good luck on your shopping endeavours!
      <br />
      <br />
      P.S. Shop responsibly!
    </Disclaimer>
  </Page>
);

export default DisclaimerPage;
