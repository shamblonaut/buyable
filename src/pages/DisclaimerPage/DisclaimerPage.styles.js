import styled from "styled-components";

export const Disclaimer = styled.p`
  padding: 32px;
  border: 3px solid var(--color-dark);
  max-width: 600px;
  border-radius: 8px;

  color: var(--color-light);
  background-color: var(--color-accent-dark);

  font-size: 1.25rem;
  font-weight: 500;
`;
