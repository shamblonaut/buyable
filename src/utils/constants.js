export const AppPosition = Object.freeze({
  HOME: 0,
  SHOP: 1,
  CART: 2,
});
