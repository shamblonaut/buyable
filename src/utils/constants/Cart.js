export const CartAction = Object.freeze({
  ADDED: "added_to_cart",
  REMOVED: "removed_from_cart",
  INCREMENTED: "incremented_cart_quantity",
  DECREMENTED: "decremented_cart_quantity",
  EDITED: "edited_cart_quantity",
  CLEARED: "cleared_cart",
});
