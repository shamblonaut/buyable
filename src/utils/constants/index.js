import { AppPosition } from "./App";

export { AppPosition };
