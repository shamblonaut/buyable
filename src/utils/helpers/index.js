import { getPositionRoute } from "./App";
import { getRatingArray } from "./Rating";
import { toTitleCase } from "./Misc";

export { getPositionRoute, getRatingArray, toTitleCase };
