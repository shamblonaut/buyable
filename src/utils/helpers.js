import { AppPosition } from "./constants";

export const getRatingArray = (rating) => {
  if (rating > 5.0) {
    throw new Error("Invalid rating");
  }

  let ratingArray = [];
  for (let i = 0; i < 5; i++) {
    const remainder = rating - i;
    ratingArray.push(
      remainder > 1 ? 1 : remainder > 0 ? Number(remainder.toFixed(1)) : 0,
    );
  }

  return ratingArray;
};

export const getPositionRoute = (appPosition) => {
  switch (appPosition) {
    case AppPosition.HOME:
      return "/";
    case AppPosition.SHOP:
      return "/shop";
    case AppPosition.CART:
      return "/cart";
  }
};

export const toTitleCase = (string) => {
  let result = "";
  for (const word of string.split(" ")) {
    result += word[0].toUpperCase() + word.substring(1) + " ";
  }
  return result.trim();
};
