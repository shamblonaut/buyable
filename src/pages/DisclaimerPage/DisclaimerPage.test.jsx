import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";

import { AppPosition } from "@/utils/constants";

import OutletWrapper from "@tests/components/OutletWrapper";

import DisclaimerPage from "./DisclaimerPage";

const mockOutletContext = {
  appPosition: AppPosition.CART,
};

describe("Disclaimer Page", () => {
  it("renders correctly", () => {
    const { container } = render(
      <OutletWrapper context={mockOutletContext}>
        <DisclaimerPage />
      </OutletWrapper>,
    );

    expect(container).toMatchSnapshot();
  });

  it("links to correct back page", () => {
    render(
      <OutletWrapper context={mockOutletContext}>
        <DisclaimerPage />
      </OutletWrapper>,
    );

    expect(screen.getByRole("link")).toHaveAttribute("href", "/cart");
  });
});
