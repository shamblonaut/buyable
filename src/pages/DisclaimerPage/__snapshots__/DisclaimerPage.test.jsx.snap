// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Disclaimer Page > renders correctly 1`] = `
<div>
  <main
    class="sc-hwkwBN kikxJF"
  >
    <a
      class="sc-fOOuSg jCktvc"
      data-discover="true"
      href="/cart"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-chevron-left"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m15 18-6-6 6-6"
        />
      </svg>
    </a>
    <p
      class="sc-kNOymR juelym"
    >
      Disclaimer
    </p>
    <p
      class="sc-hdBJTi cDFHRQ"
    >
      Sorry to disappoint you, but this was a fake online store that I built to practice React development. Good luck on your shopping endeavours!
      <br />
      <br />
      P.S. Shop responsibly!
    </p>
  </main>
</div>
`;
